spyder/api/plugins.py

```diff
@@ -533,3 +550,697 @@ def set_ancestor(self, ancestor):
+    # List of required plugin dependencies
+    REQUIRES = None
+
+    # List of optional plugin dependencies
+    OPTIONAL = None
+
+    # This is the a widget (or any non widget class) to store behavior and
+    # actions. Subclass SpyderWigdetMixin
+    WIDGET_CLASS = None
+
+    # Name of the configuration section that's going to be
+    # used to record the plugin's permanent data in Spyder
+    # config system (i.e. in spyder.ini)
+    CONF_SECTION = None
+
+    # Use separate configuration file for plugin
+    CONF_FILE = True
+
+    # Define configuration defaults if using a separate file.
+    # List of tuples, with the first item in the tuple being the section
+    # name and the second item being the default options dictionary.
+    #
+    # CONF_DEFAULTS_EXAMPLE = [
+    #     ('section-name', {'option-1': 'some-value',
+    #                       'option-2': True,}),
+    #     ('another-section-name', {'option-3': 'some-other-value',
+    #                               'option-4': [1, 2, 3],}),
+    # ]
+    CONF_DEFAULTS = None
+
+    # Define configuration version if using a separate file
+    #
+    # IMPORTANT NOTES:
+    # 1. If you want to *change* the default value of a current option, you
+    #    need to do a MINOR update in config version, e.g. from 3.0.0 to 3.1.0
+    # 2. If you want to *remove* options that are no longer needed or if you
+    #    want to *rename* options, then you need to do a MAJOR update in
+    #    version, e.g. from 3.0.0 to 4.0.0
+    # 3. You don't need to touch this value if you're just adding a new option
+    CONF_VERSION = None
+
+    # Widget to be used as entry in Spyder Preferences dialog
+    CONF_WIDGET_CLASS = None
+
+    # Path for images relative to the plugin path
+    IMG_PATH = 'images'
+
+    # Control the font size relative to the global fonts defined in Spyder
+    FONT_SIZE_DELTA = 0
+    RICH_FONT_SIZE_DELTA = 0
+
+    # --- API: Signals -------------------------------------------------------
+    # ------------------------------------------------------------------------
+
+    # Use this signal to inform another plugin that configuration changed.
+    sig_option_changed = Signal(str, object)
+    sig_toggle_view_changed = Signal(bool)
+    sig_status_message_requested = Signal(str, int)
+    sig_shortcut_registration_requested = Signal(QObject, str, str, bool)
+    sig_quit_requested = Signal()
+    sig_restart_requested = Signal()
+
+    # --- Private attributes -------------------------------------------------
+    # ------------------------------------------------------------------------
+    # Define configuration name map for plugin to split configuration
+    # among several files. See spyder/config/main.py
+    _CONF_NAME_MAP = None
+
+    # Store reference to registered plugins in a dictionary
+    # {plugin.NAME: plugin_instance, ...}
+    _PLUGINS = None
+
+    def __init__(self, parent, configuration=None):
```

Comment:
>You don't want __init__ to be overridden in subclasses? That seems quite unnatural; is that necessary?

Actually no. The idea is not to override the init, only here because we need to set some  initialization for dockable plugins on top of the "normal" plugin initialization.

The PlotsPlugin looks like:

```python
class PlotsPlugin(SpyderDockablePlugin):
    """
    Plots plugin.
    """

    NAME = 'plots'
    REQUIRES = [IPythonConsole]
    WIDGET_CLASS = PlotsWidget
    TABIFY = VariableExplorer
    CONF_SECTION = NAME  # FIXME: Redundant? move to use NAME for this as well
    CONF_FILE = False
    DISABLE_ACTIONS_WHEN_HIDDEN = False

    def get_name(self):
        return _('Plots')

    def get_description(self):
        return _('Display, explore and save console generated plots.')

    def get_icon(self):
        return self.create_icon('hist')

    def register(self):
        # Plugins
        ipyconsole = self.get_plugin(IPythonConsole)

        # Signals
        widget = self.get_widget()
        ipyconsole.sig_shellwidget_id_changed.connect(
            widget.set_shellwidget_from_id)
        ipyconsole.sig_shellwidget_process_started.connect(
            widget.add_shellwidget)
        ipyconsole.sig_shellwidget_id_process_finished.connect(
            widget.remove_shellwidget_from_id)

    # Exposed API
    def current_widget(self):
        return self.get_widget().current_widget()
```